### PR TITLE
Implement NodeCreationExceptions for Structured Error Handling

### PR DESCRIPTION
**Pull request by Aryan-Railtown**
## Summary
Implements the first phase of structured exception handling for the RC framework by adding comprehensive node creation validation with detailed, formatted error messages.

### Core Implementation
- **New validation functions** in `exceptions/node_creation/validators.py`:
  - `check_classmethod()` - Validates method is a classmethod
  - `check_connected_nodes()` - Ensures node connections are valid
  - `check_duplicate_param_names()` - Prevents duplicate parameter names
  - `check_output_model()` - Validates Pydantic output models
  - `check_pretty_name()` - Ensures pretty names are provided when needed
  - `check_system_message()` - Validates system message types
  - `check_tool_params_and_details()` - Ensures tool consistency

### Exception Messaging System
- **Centralized message management** via `exception_messages.yaml`
- **Structured error format** with main messages and detailed notes
- **Template support** for dynamic error information
- **Consistent formatting** across all node creation errors

### Error Handling Integration
- **NodeCreationError** integration from existing fatal exception system
- **Color-coded output** for enhanced error visibility
- **Detailed validation** at node creation time
- **Actionable error messages** with specific guidance

### Testing
- **Comprehensive unit tests** covering all validation scenarios
- **Edge case coverage** for malformed inputs
- **Exception message verification** ensuring correct error details

## Benefits
- ✅ Clear, actionable error messages for node creation failures
- ✅ Centralized exception message management
- ✅ Backward compatibility maintained
- ✅ Foundation laid for additional exception types (NodeInvocationExceptions next)
- ✅ Improved developer experience with detailed validation feedback

## Next Steps
This PR establishes the foundation for structured exception handling. The next phase will implement `NodeInvocationExceptions` for runtime error handling.

## Checklist for Author

- [x] Linked your tickets to this pull request
- [x] Ran tests locally and all pass

---

### CODE REVIEW

File: src/requestcompletion/exceptions/base.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    """
+    A simple base class for all RCExceptions to inherit from.
+    """
+
+    RED = "\033[91m"
+    GREEN = "\033[92m"
+    RESET = "\033[0m"
+
+    @classmethod
+    def color(cls, text, color_code):
```
Comment (soulFood5632) [inline src/requestcompletion/exceptions/base.py]: Probably should be a private method. `_color`

File: src/requestcompletion/exceptions/node_creation/exception_messages.yaml
Comment (soulFood5632) [inline src/requestcompletion/exceptions/node_creation/exception_messages.yaml:1]: Very clean. This will be a great model for how we do logging.

File: src/requestcompletion/exceptions/node_creation/exception_messages.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  - "Add @classmethod decorator to '{method_name}'."
+  - "Signature should be: \n@classmethod\ndef {method_name}(cls): ..."
+
+CONNECTED_NODES_EMPTY_MSG: "connected_nodes must not return an empty set."
+CONNECTED_NODES_EMPTY_NOTES:
+  - "Please provide a set nodes that can be used as tools by the ToolCallLLM node."
+
+CONNECTED_NODES_TYPE_MSG: "connected_nodes must return a set of Nodes."
+CONNECTED_NODES_TYPE_NOTES:
+  - "Ensure all the nodes provided as connected_nodes are of type Node."
```
Comment (soulFood5632) [inline src/requestcompletion/exceptions/node_creation/exception_messages.yaml]: See #160 but this is gonna change.

File: src/requestcompletion/exceptions/node_creation/exception_messages.yaml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  - "If you have functions that you want to use as tools, please use the from_function method to convert them to Nodes."
+
+DUPLICATE_PARAMETER_NAMES_MSG: "Duplicate parameter names are not allowed."
+DUPLICATE_PARAMETER_NAMES_NOTES:
+  - "Parameter names in tool_params must be unique."
+
+INVALID_SYSTEM_MESSAGE_MSG: "system_message must be a SystemMessage object, not a string or any other type."
```
Comment (soulFood5632) [inline src/requestcompletion/exceptions/node_creation/exception_messages.yaml:18]: This is a good catch for now. We will want to allow string type objects in a issue I need to create.

File: src/requestcompletion/exceptions/node_creation/validation.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+    Raises:
+        NodeCreationError: If node_set is empty or contains invalid types.
+    """
+    if not node_set:
+        raise NodeCreationError(
+            message=get_message(ExceptionMessageKey.CONNECTED_NODES_EMPTY_MSG),
+            notes=get_notes(ExceptionMessageKey.CONNECTED_NODES_EMPTY_NOTES),
+        )
+    elif not all(issubclass(x, node) for x in node_set):    # ideally we should be importing node from requestcompletion, but we don't want to deal woth circular imports here :'
```
Comment (soulFood5632) [inline src/requestcompletion/exceptions/node_creation/validation.py]: I don't think there is a good solution here, but yeah its quite annoying.

File: src/requestcompletion/nodes/library/easy_usage_wrappers/structured_llm.py [hunk 5]
```diff
@@ -87,21 +92,9 @@ def prepare_tool(cls, tool_parameters: Dict[str, Any]) -> Self:
...
-        )
-    elif (
-        tool_details
-        and tool_params
-        and len({param.name for param in tool_params}) != len(tool_params)
-    ):
-        raise ValueError("Duplicate parameter names are not allowed.")
-    if not output_model or len(output_model.model_fields) == 0:
-        raise ValueError("Output model cannot be empty")
+    check_tool_params_and_details(tool_params, tool_details)
```
Comment (soulFood5632) [inline src/requestcompletion/nodes/library/easy_usage_wrappers/structured_llm.py]: Is there anyway to group these tasks in a `validate_params` functions so we don't have repeated code.

File: src/requestcompletion/nodes/nodes.py [hunk 6]
```diff
@@ -42,6 +40,31 @@ async def async_wrapper(self, *args, **kwargs):
...
+            method = dct["output_model"]
+            check_classmethod(method, "output_model")
+            check_output_model(method, cls)
+
+        # 3. Check if the connected_nodes is not empty, special case for ToolCallLLM
+        if "connected_nodes" in dct and not getattr(cls, "__abstractmethods__", False):
+            method = dct["connected_nodes"]
+            try:  # in case of class based init
+                node_set = method.__func__(cls)
+            except AttributeError:  # in case of easy_wrapper init
```
Comment (soulFood5632) [inline src/requestcompletion/nodes/nodes.py]: If you can comment this logic a bit more that would be great.

File: src/requestcompletion/pubsub/publisher.py [hunk 7]
```diff
@@ -35,8 +35,8 @@ async def trigger(self, message: _T):
             result = self.callback(message)
             if asyncio.iscoroutine(result):
                 await result
-        except Exception as e:
-            logger.exception(msg=f"Error in {self.name}", exc_info=e)
+        except Exception:
+            pass
```
Comment (soulFood5632) [inline src/requestcompletion/pubsub/publisher.py]: Make sure we flag and deal with this during logging.

File: src/requestcompletion/state/state.py [hunk 8]
```diff
@@ -363,21 +363,25 @@ async def _handle_failed_request(
...
 
         # fatal exceptions should only be thrown if there is something seriously wrong.
         if isinstance(exception, FatalError):
             self.logger.critical(node_exception_action.to_logging_msg())
-            ee = ExecutionError(
-                failed_request=self._request_heap[request_id],
-                execution_info=self.info,
-                final_exception=exception,
+            ee = NodeInvocationError(
+                message=node_exception_action.to_logging_msg(),
```
Comment (soulFood5632) [inline src/requestcompletion/state/state.py:376]: I know this will probably change, but the notes don't make sense here. The system is shutting down due to the throwing a fatal error not the end on error flag.

File: tests/rc_tests/interaction/conftest.py [hunk 9]
```diff
@@ -287,6 +287,7 @@ def __init__(
                     message_history.insert(0, system_message)
                     super().__init__(message_history=message_history, model=llm_model)
 
+                @classmethod
                 def output_model(self) -> BaseModel:
```
Comment (soulFood5632) [inline tests/rc_tests/interaction/conftest.py]: Should be changed to `cls` not `self`

File: src/requestcompletion/state/state.py [hunk 10]
```diff
@@ -363,21 +363,25 @@ async def _handle_failed_request(
...
+                notes=[
+                    "This is a fatal exception. The system will be shutting down.",
+                    "If you want to continue running the system, you can set the `end_on_error` flag to `False` in the executor config.",
```
Comment (soulFood5632) [inline src/requestcompletion/state/state.py]: Lets go a bit deeper on this logic. I'm thinking it would be a good idea to separate these types of errors where they are caught and then thrown into a `NodeInvocationError`. We will probably want throw the actual error that caused the failure instead of wrapping into our own internal error. There is still a need for NodeInvocationErrors but I am not sure if this place makes sense.

Comment (soulFood5632) [reply src/requestcompletion/state/state.py]: Let me know if that comment makes sense. We may need a further discussion.

File: src/requestcompletion/exceptions/node_creation/exception_messages.yaml (hunk 2 above)
Comment (Aryan-Railtown) [reply src/requestcompletion/exceptions/node_creation/exception_messages.yaml]: I see, I will modify the test cases accordingly.

File: src/requestcompletion/exceptions/node_creation/exception_messages.yaml (hunk 3 above)
Comment (Aryan-Railtown) [reply src/requestcompletion/exceptions/node_creation/exception_messages.yaml:18]: Sounds good. I will leave this in the meantime.

File: src/requestcompletion/nodes/library/easy_usage_wrappers/structured_llm.py (hunk 5 above)
Comment (Aryan-Railtown) [reply src/requestcompletion/nodes/library/easy_usage_wrappers/structured_llm.py]: Good catch, we ideally should, since these 4 checks will be done for all easy_usage wrappers.

File: src/requestcompletion/state/state.py (hunk 10 above)
Comment (Aryan-Railtown) [reply src/requestcompletion/state/state.py]: I see. That kind of makes sense, but I will reach out to discuss more about this once I get to this. For this task I was just refactoring the pre-existing ExecutionError, so I replaaced it with the NodeInvocationError. I can flag this area for now and remove the notes here.

File: src/requestcompletion/exceptions/node_creation/exception_messages.yaml (hunk 3 above)
Comment (soulFood5632) [reply src/requestcompletion/exceptions/node_creation/exception_messages.yaml:18]: Issues created 👍
